aten/src/ATen/native/sparse/SparseFactories.cpp

```diff
@@ -0,0 +1,126 @@
+#include <ATen/Dispatch.h>
+#include <ATen/SparseTensorImpl.h>
+#include <ATen/SparseTensorUtils.h>
+#include <ATen/TensorIndexing.h>
+#include <ATen/TensorIterator.h>
+#include <ATen/core/ATen_fwd.h>
+#include <ATen/core/Tensor.h>
+#include <ATen/native/cpu/Loops.h>
+#include <ATen/native/sparse/SparseFactories.h>
+#include <c10/core/Scalar.h>
+#include <c10/util/ArrayRef.h>
+#include <c10/util/Exception.h>
+
+#ifndef AT_PER_OPERATOR_HEADERS
+#include <ATen/Functions.h>
+#include <ATen/NativeFunctions.h>
+#else
+#include <ATen/ops/sparse_coo_tensor.h>
+#endif
+
+namespace at {
+namespace native {
+using namespace at::sparse;
+/******************************************************************************
+ * Build sparse from diagonals
+ ******************************************************************************/
+
+// --------------------------------------------------------------------
+// spdiags(D, O, (N,M)) -> S
+//
+// Take rows of D and place them on the diagonals specified by offsets O of a
+// new (NxM) sparse matrix S If D is (P x Q) then O must be a row vector (P, ).
+// It does not matter if Q values fit  on any diagonal of S, or if S has no
+// O[i]th diagonal (those values/diagonals are simply skipped)
+// --------------------------------------------------------------------
+
+namespace {
+void _spdiags_kernel_cpu(
+    TensorIterator& iter,
+    const Tensor& diagonals,
+    Tensor& values,
+    Tensor& indices) {
+  auto* row_index_write_ptr = indices[0].data_ptr<int64_t>();
+  auto* col_index_write_ptr = indices[1].data_ptr<int64_t>();
+  const int64_t diagonals_read_stride = diagonals.stride(1);
+  AT_DISPATCH_ALL_TYPES_AND_COMPLEX_AND4(
+      at::ScalarType::BFloat16,
+      at::ScalarType::Half,
+      at::ScalarType::Bool,
+      at::ScalarType::ComplexHalf,
+      diagonals.scalar_type(),
+      "spdiags_cpu",
+      [&] {
+        auto* values_write_ptr = values.data_ptr<scalar_t>();
+        cpu_kernel(
+            iter,
+            [&](int64_t diag_index,
+                int64_t diag_offset,
+                int64_t out_offset,
+                int64_t n_out) -> int64_t {
+              auto* rows_start = row_index_write_ptr + out_offset;
+              auto* cols_start = col_index_write_ptr + out_offset;
+              auto* vals_start = values_write_ptr + out_offset;
+              int64_t const first_col = std::max<int64_t>(diag_offset, 0);
+              int64_t const first_row = first_col - diag_offset;
+              auto* data_read = diagonals[diag_index].data_ptr<scalar_t>() +
+                  first_col * diagonals_read_stride;
```

Comment:
We need to know which diagonal we are working with so we know which row of `diagonals` to read from. I can either re-stride diagonals so that its pointer will only advance along the `n_diag` dimension, or provide the `arange(n_diag)` as input.